zuul-discovery/src/main/java/com/netflix/zuul/discovery/DynamicServerResolver.java

```diff
@@ -0,0 +1,95 @@
+package com.netflix.zuul.discovery;
+
+import static com.netflix.client.config.CommonClientConfigKey.NFLoadBalancerClassName;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Sets;
+import com.netflix.client.config.IClientConfig;
+import com.netflix.loadbalancer.DynamicServerListLoadBalancer;
+import com.netflix.loadbalancer.Server;
+import com.netflix.niws.loadbalancer.DiscoveryEnabledServer;
+import com.netflix.zuul.resolver.Resolver;
+import com.netflix.zuul.resolver.ResolverListener;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * @author Argha C
+ * @since 2/25/21
+ *
+ * Implements a resolver, wrapping a ribbon load-balancer.
+ */
+public class DynamicServerResolver implements Resolver<DiscoveryResult> {
+
+    private final DynamicServerListLoadBalancer<?> loadBalancer;
+    ResolverListener<DiscoveryResult> listener;
+
+    public DynamicServerResolver(IClientConfig clientConfig, ResolverListener<DiscoveryResult> listener) {
+        this.loadBalancer = createLoadBalancer(clientConfig);
+        this.loadBalancer.addServerListChangeListener(this::onUpdate);
+        this.listener = listener;
+    }
+
+    @Override
+    public DiscoveryResult resolve(@Nullable Object key) {
+        final Server server = loadBalancer.chooseServer(key);
+        return new DiscoveryResult((DiscoveryEnabledServer) server);
+    }
+
+    public boolean hasServers() {
+        return !loadBalancer.getReachableServers().isEmpty();
+    }
+
+    public void shutdown() {
+        loadBalancer.shutdown();
+    }
+
+    private DynamicServerListLoadBalancer<?> createLoadBalancer(IClientConfig clientConfig) {
+        //TODO(argha-c): Revisit this style of LB initialization post modularization. Ideally the LB should be pluggable.
+
+        // Use a hard coded string for the LB default name to avoid a dependency on Ribbon classes.
+        String loadBalancerClassName =
+                clientConfig.get(NFLoadBalancerClassName, "com.netflix.loadbalancer.ZoneAwareLoadBalancer");
+
+        DynamicServerListLoadBalancer<?> lb;
+        try {
+            Class<?> clazz = Class.forName(loadBalancerClassName);
+            lb = clazz.asSubclass(DynamicServerListLoadBalancer.class).getConstructor().newInstance();
+            lb.initWithNiwsConfig(clientConfig);
+        } catch (Exception e) {
+            Throwables.throwIfUnchecked(e);
+            throw new IllegalStateException("Could not instantiate LoadBalancer " + loadBalancerClassName, e);
+        }
+
+        return lb;
+    }
+
+    @VisibleForTesting
+    void onUpdate(List<Server> oldList, List<Server> newList) {
+        Set<Server> oldSet = new HashSet<>(oldList);
+        Set<Server> newSet = new HashSet<>(newList);
+        final List<DiscoveryResult> discoveryResults = Sets.difference(oldSet, newSet).stream()
```

Comment:
Shouldn't this be the symmetric difference between the two?

Reply:
The listener interface is designed to only receive the set that was _removed_ from the last update. In effect, that isn't symmetric, since we're only interested in `old - new`.
Arguable if that's the best design since I tried to fit it to the current implementation.